src/boolean3.cpp

```diff
@@ -448,44 +337,137 @@ struct Kernel12 {
       v12.y = xzyy[2];
       v12.z = xzyy[1];
     }
+    return std::make_pair(x12, v12);
   }
 };
 
-std::tuple<Vec<int>, Vec<vec3>> Intersect12(
-    const Manifold::Impl &inP, const Manifold::Impl &inQ, const Vec<int> &s02,
-    const SparseIndices &p0q2, const Vec<int> &s11, const SparseIndices &p1q1,
-    const Vec<double> &z02, const Vec<vec4> &xyzz11, SparseIndices &p1q2,
-    bool forward) {
+struct Kernel12Tmp {
+  SparseIndices p1q2_;
+  Vec<int> x12_;
+  Vec<vec3> v12_;
+};
+
+struct Kernel12Recorder {
+  using LocalT = Kernel12Tmp;
+  Kernel12 &k12;
+  VecView<const TmpEdge> tmpedges;
+  bool forward;
+
+#if MANIFOLD_PAR == 1
+  tbb::combinable<Kernel12Tmp> store;
+  LocalT &local() { return store.local(); }
+#else
+  Kernel12Tmp localStore;
+  LocalT &local() { return localStore; }
+#endif
+
+  void record(int queryIdx, int leafIdx, LocalT &tmp) {
+    queryIdx = tmpedges[queryIdx].halfedgeIdx;
+    const auto [x12, v12] = k12(queryIdx, leafIdx);
+    if (std::isfinite(v12[0])) {
+      if (forward)
+        tmp.p1q2_.Add(queryIdx, leafIdx);
+      else
+        tmp.p1q2_.Add(leafIdx, queryIdx);
+      tmp.x12_.push_back(x12);
+      tmp.v12_.push_back(v12);
+    }
+  }
+
+  Kernel12Tmp get() {
+#if MANIFOLD_PAR == 1
+    Kernel12Tmp result;
+    std::vector<Kernel12Tmp> tmps;
+    std::vector<SparseIndices> indices;
+    store.combine_each([&](Kernel12Tmp &data) {
+      indices.emplace_back(std::move(data.p1q2_));
+      tmps.emplace_back(std::move(data));
+    });
+    result.p1q2_.Clear();
+    result.p1q2_.FromIndices(indices);
+
+    std::vector<size_t> sizes;
+    size_t total_size = 0;
+    for (const auto &tmp : tmps) {
+      sizes.push_back(total_size);
+      total_size += tmp.x12_.size();
+    }
+    result.x12_.resize(total_size);
+    result.v12_.resize(total_size);
+    for_each_n(ExecutionPolicy::Seq, countAt(0), indices.size(), [&](size_t i) {
+      std::copy(tmps[i].x12_.begin(), tmps[i].x12_.end(),
+                result.x12_.begin() + sizes[i]);
+      std::copy(tmps[i].v12_.begin(), tmps[i].v12_.end(),
+                result.v12_.begin() + sizes[i]);
+    });
+    return result;
+#else
+    return localStore;
+#endif
+  }
+};
+
+std::tuple<Vec<int>, Vec<vec3>> Intersect12(const Manifold::Impl &inP,
+                                            const Manifold::Impl &inQ,
+                                            SparseIndices &p1q2, double expandP,
+                                            bool forward) {
   ZoneScoped;
-  Vec<int> x12(p1q2.size());
-  Vec<vec3> v12(p1q2.size());
+  // a: 1 (edge), b: 2 (face)
+  const Manifold::Impl &a = forward ? inP : inQ;
```

Comment:
Now, instead of swapping `inP` and `inQ` parameters, I just define new temporaries a and b, depending on `forward`. This should hopefully be cleaner now.